### DISCUSSION TITLE
Kerberos SSO Support in FreeRDP Windows Client

### DISCUSSION THREAD

**Discussion by rofferom** [FreeRDP/FreeRDP#10756] [Q&A]
Hi,

I have a question regarding the current implementation status of Kerberos and SSO support in FreeRDP. As I'm relatively new to both RDP and Kerberos, I'm working to build my understanding of these technologies, so some of my observations might be incorrect.

## Use case

To start, I'd like to outline my target environment and use case. My goal is to use the FreeRDP client on Windows to connect to a Windows Server, with Kerberos as the authentication backend (both machines are in the same domain). Ultimately, I want to enable Single Sign-On (SSO) so that the user only logs in once.

Here’s an example of what I aim to achieve:
1. The user logs into their physical PC
2. The user opens a remote session on a server without re-entering their credentials
3. The user is connected to a remote session where all actions are performed on their behalf, with SSO also enabled. This would mean the user would never retype their credentials during the remote session.

## FreeRDP-Specific questions

From what I understand, this use case is possible in a fully Windows-based environment using the official RDP client, with some specific GPOs on the Active Directory side.

Although FreeRDP implements Kerberos support, most discussions seem to focus on the Linux (and possibly macOS) versions. Therefore, I would like to ask about the implementation status for the FreeRDP Windows client:

Is it currently possible to achieve the use case described above using FreeRDP on Windows?

If not, what features/blockers are missing?

## Technical Questions and Observations

I've done some research into the RDP authentication process, particularly in relation to Kerberos.

Regarding Kerberos, assuming the correct GPOs are configured, I understand from [this MSDN page](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-sfu/1fb9caca-449f-4183-8f7a-1a5fc7e7290a) that it's possible to use the TGT obtained during the initial login (step 1) to generate a TGS for the RDS server. 

When reviewing documentation about CredSSP, it seems the only method to pass a TGS through `TSCredentials` would be to use `TSRemoteGuardCreds`. Both `TSPasswordCreds` and `TSSmartCardCreds` carry the user credentials, which isn't what is expected when using Kerberos.

Also, the MSDN documentation states regarding `TSRemoteGuardCreds`:
> The TSRemoteGuardCreds structure SHOULD contain a logon credential and supplemental credentials provided by security packages. The format of the individual credentials depends on the package that provided them.

My understanding is that this message will carry the output of SPPI's `InitializeSecurityContext()` `AcquireCredentialsHandle()` call sequence, which would be the TGS?

A [rdesktop github issue](https://github.com/rdesktop/rdesktop/issues/401) mentions Remote Credential Guard is the proper way to forward Kerberos tickets, thereby enabling SSO.

Here is my question: are my CredSSP observation regarding `TSCredentials` and the rdesktop github issue valid? Is Remote Credential Guard the only way to make proper Kerberos authentication (i.e., without sending client credentials to the server) with SSO support on RDP?

If true, #10549 seems to implement the client side for Remote Credential Guard. But the PR notes that:
> For now it's UNIX only as the windows implementation would beimplemented in a completely different way

Could you confirm whether any features are still missing to make this use case work on Windows?

**Comment by hardening** [comment]
Hi,

so under windows we'd need to use the same approach as mstsc to achieve what you want. That is using the tspkg SSPI module that would do NLA for us in lsass and so would have access to all session's passwords / NTLM hashes / Kerberos ccache. As we're doing by ourself all of that in FreeRDP, we can only provide SSO with things that we have: so credentials given on the command line.
So even when you're doing kerberos in NLA, you need to provide credentials so that SSO works on the remote server (except if you choosed restricted admin). When lsass forges your nla packets (as does mstsc by using the tspkg SSPI module) it's rather easy (as lsass knows your credentials) to build that `TSPasswordCreds`.

For now RCG is not implemented on windows more or less for this reason: the channel would just pass the messages to LSASS, so quite easy to implement. But to connect with RCG we'd need to use TSRemoteGuardCreds in NLA, and so provide the TGT and a TGS, I don't know how to do that by requesting kerberos dll. Mstsc uses the tspkg SSPI module so it's basically lsass that does all these kerberos stuff in NLA packets. I used to do a bit of RE on tspkg.dll, but there's lots of thing there. And there's no magic when you do things in lsass you have access to all the secrets associated with the session, so the best place for SSO. 

Perhaps we should seriously consider not doing NLA by ourself on windows, but that would require a massive refactoring of this part of FreeRDP. Perhaps we should write a `tspkg` SSPI module in winpr for non windows and move NLA code there, so this way that would work the same on all OSs. In fact that's a little what @fifthdegree did when he abstracted some part of NLA in `libfreerdp/core/credssp_auth`...

**Reply by rofferom** [reply]
Thanks for you feedback.

I have some extra questions to ensure my understanding of some details of RDP and Kerberos interaction:
1. What I understood about Nego part of CredSSP: TGS is acquired by client and sent to server here. However, when TGT is sent? Do yo mean that when we are using NLA without RCG, we still need to provide `TSPasswordCreds` for the server to obtain a TGT on behalf of the client?
2. Is "Restricted admin" mode enabled when an empty `TSPasswordCreds` is sent?
3. I found few documentation about `tspkg`, but you mentioned you've done some research on it. Regardless of the refactoring required for FreeRDP, do you think `tspkg` integration is likely to work, or is it still mostly a big black box?

**Reply by hardening** [reply]
> Thanks for you feedback.
> 
> I have some extra questions to ensure my understanding of some details of RDP and Kerberos interaction:
> 
>     1. What I understood about Nego part of CredSSP: TGS is acquired by client and sent to server here. However, when TGT is sent? Do yo mean that when we are using NLA without RCG, we still need to provide `TSPasswordCreds` for the server to obtain a TGT on behalf of the client?
> 

So when doing RCG, the implicit is that you're doing Kerberos authentication in SPNego, or at least any kind of authentication that provides Kerberos ticket. I'd say that it allows to pass the authentication step of NLA. At the end of NLA done with Kerberos, you can either send `TSPasswordCreds` for regular user/password SSO (you can send them empty if you have set the restricted admin flag), or you can send `TSRemoteGuardCreds` indicating that you wanna do RCG. RCG means that basically the remote server will remote all the LSASS call to the RDP client. So in the remote session, as soon as you trigger something that would require credentials stored in LSASS, a message is sent to the RDP client to serve the requested item.

>     2. Is "Restricted admin" mode enabled when an empty `TSPasswordCreds` is sent?
> 
Not only, you must also set a flag in the first nego packet.

>     3. I found few documentation about `tspkg`, but you mentioned you've done some research on it. Regardless of the refactoring required for FreeRDP, do you think `tspkg` integration is likely to work, or is it still mostly a big black box?

In theory it should work on windows as we would do exactly like `mstsc`. For some cases as with childSession connection that would also allow to have SSO working (right now you must pass the user/password to FreeRDP where with mstsc you don't need them as it just pass some magic value in SSPI calls and lsass knows that it's trusted childSession connection).

Yeah `tspkg` it's not documented, but we know what it's doing.

**Reply by fifthdegree** [reply]
@hardening I remember playing with LsaCallAuthenticationPackage to retrieve tickets IIRC there's a registry setting that needs to be set for windows to give you the TGT session key

**Reply by hardening** [reply]
@fifthdegree hum but when you use `mstsc` there's no such registry key trick. And we'd need the TGT but also a TGS...
